core/context.go

```diff
@@ -454,3 +508,58 @@ func (ctx *Context) GetBackend(vsID, rsID string) (*BackendInfo, error) {
 
 	return &BackendInfo{rs.options, rs.metrics}, nil
 }
+
+// if external kvstore exists, set store to context
+func (ctx *Context) SetStore(store *Store) {
+	ctx.store = store
+}
+
+func (ctx *Context) Synchronize(storeServices map[string]*ServiceOptions, storeBackends map[string]*BackendOptions) {
+	ctx.mutex.Lock()
+	defer ctx.mutex.Unlock()
+
+	log.Debugf("============================== SYNC ========================================")
+	for k, v := range storeServices {
+		log.Debugf("SERVICE[%s]: %s", k, v)
+	}
+	for k, v := range storeBackends {
+		log.Debugf("  BACKEND[%s]: %s", k, v)
+	}
+	defer log.Debugf("============================================================================")
+
+	// synchronize services with store
+	for id, _ := range ctx.services {
+		if _, ok := storeServices[id]; !ok {
+			ctx.removeService(id)
+		}
+	}
+	for id, storeServiceOptions := range storeServices {
+		if service, ok := ctx.services[id]; ok {
+			if service.options.CompareStoreOptions(storeServiceOptions) {
+				continue
+			}
+			ctx.removeService(id)
+		}
+		ctx.createService(id, storeServiceOptions)
+	}
+
+	// synchronize backends with store
+	for id, backend := range ctx.backends {
+		if _, ok := storeBackends[id]; !ok {
+			vsID := "(unknown)"
```

Comment:
I agree that we shouldn't update any IPVS configuration if KV store update has failed. I don't think we should exit Gorb is 1st sync has failed though - just keep trying.